### PR TITLE
Fixes #25037 - Add CV import command

### PR DESCRIPTION
**Pull request by ehelms**

### CODE REVIEW

**Comment by ehelms** [comment]
Currently depends upon https://github.com/Katello/hammer-cli-katello/pull/589 and should be merged afterward. This is dependent upon the final output of the of the export JSON and will require https://github.com/Katello/katello/pull/7696 to be merged before all functionality is viable. I am opening this for testing purposes and to refine the workflow.

File: lib/hammer_cli_katello/content_view_version.rb [hunk 1]
```diff
@@ -259,6 +259,208 @@ class ExportCommand < HammerCLIKatello::SingleResourceCommand
...
+        repo_units = @export_json['repositories'].collect { |repo| repo['rpm_filenames'] }.uniq
+        publish(cv['id'], @export_json['major'], @export_json['minor'], repo_units)
+      end
+
+      def untar_export
+        Dir.chdir(@export_tar_dir) do
+          `tar -xf #{@export_tar_file}`
+
+          if File.exist?("#{@export_tar_file.gsub('.tar', '-repos.tar')}")
+            `tar -xf #{@export_tar_file.gsub('.tar', '-repos.tar')}`
```
Comment (chris1984) [inline lib/hammer_cli_katello/content_view_version.rb]: On export it looks like we are getting exported-repos-<id>.tar which causes this to fail:
~~~
exported-repos-3.tar
~~~

~~~
[root@import ~]# ls
anaconda-ks.cfg  export-3.json  export-3.tar  exported-repos-3.tar  ssl-build
hammer> content-view version import --export-tar /root/exported-repos-3.tar
Could not import the content view:
  Error: Export repos tar file is missing
[root@import ~]# mv exported-repos-3.tar exported-repos.tar 
hammer> content-view version import --export-tar /root/exported-repos.tar
Could not import the content view:
  Error: Export repos tar file is missing
hammer> content-view version import --export-tar /root/exported-repos.tar
[root@import ~]# mv exported-repos.tar repos.tar
[root@import ~]# ls
export-3.json  export-3.tar  repos.tar  ssl-build
[root@import ~]# hammer shell
Welcome to the hammer interactive shell.
Type 'help' for usage information.
hammer> content-view version import --export-tar /root/repos.tar
Could not import the content view:
  Error: Export repos tar file is missing
hammer> content-view version import --export-tar /root/export-3.tar
Could not import the content view:
  Error: Export repos tar file is missing
~~~
Tried renaming it to a few things and its not seeing it, selinux is off

Comment (ehelms) [reply lib/hammer_cli_katello/content_view_version.rb]: I tried to clean this up. I think they should be consistent now.

**Comment by chris1984** [comment]
@ehelms getting past the tar error, but hitting a different one:
~~~
[root@import ~]# hammer -d content-view version import --export-tar /root/export-3.tar
[ INFO 2018-09-27T14:51:12 Init] Initialization of Hammer CLI (0.14.0) has started...
[DEBUG 2018-09-27T14:51:12 Init] Running at ruby 2.4.3-p205
[ INFO 2018-09-27T14:51:12 Init] Configuration from the file /etc/hammer/cli_config.yml has been loaded
[ INFO 2018-09-27T14:51:12 Init] Configuration from the file /etc/hammer/cli.modules.d/foreman.yml has been loaded
[ INFO 2018-09-27T14:51:12 Init] Configuration from the file /etc/hammer/cli.modules.d/foreman_bootdisk.yml has been loaded
[ INFO 2018-09-27T14:51:12 Init] Configuration from the file /etc/hammer/cli.modules.d/foreman_docker.yml has been loaded
[ INFO 2018-09-27T14:51:12 Init] Configuration from the file /etc/hammer/cli.modules.d/foreman_tasks.yml has been loaded
[ INFO 2018-09-27T14:51:12 Init] Configuration from the file /etc/hammer/cli.modules.d/katello.yml has been loaded
[ INFO 2018-09-27T14:51:12 Init] Configuration from the file /root/.hammer/cli.modules.d/foreman.yml has been loaded
[DEBUG 2018-09-27T14:51:12 SSLoptions] SSL options: {
    :ssl_ca_file => "/etc/pki/katello/certs/katello-server-ca.crt",
     :verify_ssl => true
}
[DEBUG 2018-09-27T14:51:12 API] Global headers: {
        :content_type => "application/json",
              :accept => "application/json;version=2",
    "Accept-Language" => "en"
}
[DEBUG 2018-09-27T14:51:12 API] Follow redirects: never
[DEBUG 2018-09-27T14:51:12 Connection] Registered: foreman
[ INFO 2018-09-27T14:51:12 Modules] Extension module hammer_cli_foreman (0.14.0) loaded.
[ INFO 2018-09-27T14:51:12 Modules] Extension module hammer_cli_foreman_bootdisk (0.1.3) loaded.
[ INFO 2018-09-27T14:51:12 HammerCLI::MainCommand] subcommand organization (HammerCLIForeman::Organization) was removed.
[ INFO 2018-09-27T14:51:12 HammerCLI::MainCommand] subcommand organization (HammerCLIKatello::Organization) was created.
[ INFO 2018-09-27T14:51:12 Modules] Extension module hammer_cli_katello (0.14.1) loaded.
[DEBUG 2018-09-27T14:51:12 Init] Using locale 'en'
[DEBUG 2018-09-27T14:51:12 Init] 'mo' files for locale domain 'hammer-cli' loaded from '/opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.14.0/locale'
[DEBUG 2018-09-27T14:51:12 Init] 'mo' files for locale domain 'hammer-cli-foreman' loaded from '/opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli_foreman-0.14.0/locale'
[DEBUG 2018-09-27T14:51:12 Init] 'mo' files for locale domain 'hammer_cli_foreman_docker' loaded from '/opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli_foreman_docker-0.0.4/locale'
[DEBUG 2018-09-27T14:51:12 Init] 'mo' files for locale domain 'hammer-cli-katello' loaded from '/opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli_katello-0.14.1/locale'
[ INFO 2018-09-27T14:51:12 HammerCLI::MainCommand] Called with options: {"option_debug"=>true}
[ INFO 2018-09-27T14:51:13 HammerCLIKatello::ContentView] Called with options: {}
[ INFO 2018-09-27T14:51:13 HammerCLIKatello::ContentViewVersion] Called with options: {}
[ INFO 2018-09-27T14:51:13 HammerCLIKatello::ContentViewVersion::ImportCommand] Called with options: {"option_export_tar"=>"/root/export-3.tar"}
[DEBUG 2018-09-27T14:51:13 Exception] Using exception handler HammerCLIKatello::ExceptionHandler#handle_general_exception
[ERROR 2018-09-27T14:51:13 Exception] Error: undefined method `[]' for nil:NilClass
Could not import the content view:
  Error: undefined method `[]' for nil:NilClass
[ERROR 2018-09-27T14:51:13 Exception] 

NoMethodError (undefined method `[]' for nil:NilClass):
    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli_katello-0.14.1/lib/hammer_cli_katello/content_view_version.rb:619:in `execute'
    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.14.0/lib/hammer_cli/abstract.rb:29:in `run'
    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.14.0/lib/hammer_cli/abstract.rb:29:in `run'
    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.14.0/lib/hammer_cli/abstract.rb:29:in `run'
    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.14.0/lib/hammer_cli/abstract.rb:29:in `run'
    /opt/theforeman/tfm/root/usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:132:in `run'
    /opt/theforeman/tfm/root/usr/share/gems/gems/hammer_cli-0.14.0/bin/hammer:143:in `<top (required)>'
    /usr/bin/hammer:23:in `load'
    /usr/bin/hammer:23:in `<main>'
~~~

File: lib/hammer_cli_katello/content_view_version.rb [hunk 2]
```diff
@@ -259,6 +261,214 @@ class ExportCommand < HammerCLIKatello::SingleResourceCommand
...
+      end
+
+      def sync_repositories
+        @export_json['repositories'].each do |repo|
+          synchronize(repo['id'], "file://#{@export_tar_dir}/#{@export_tar_prefix}/#{repo['relative_path']}")
+        end
+      end
+
+      def content_view(name)
+        HammerCLIForeman.foreman_resource(:content_views).call(:index, 'name' => name)['results'].first
```
Comment (akofink) [inline lib/hammer_cli_katello/content_view_version.rb]: Silently fails if you have multiple orgs with the same content view names :(

File: lib/hammer_cli_katello/content_view_version.rb (hunk 1 above)
Comment (beav) [reply lib/hammer_cli_katello/content_view_version.rb]: this worked for me, :+1: to resolve comment

File: lib/hammer_cli_katello/content_view_version.rb [hunk 3]
```diff
@@ -385,6 +385,130 @@ def export_json(content_view_version, repositories)
...
+        cv = content_view(export_json['name'])
+        sync_repositories(export_json['repositories'])
+
+        publish(
+          cv['id'],
+          export_json['major'],
+          export_json['minor'],
+          repos_units(export_json['repositories'])
+        )
+        return HammerCLI::EX_OK
```
Comment (beav) [inline lib/hammer_cli_katello/content_view_version.rb]: was this just added? I was about to request it be added but saw it was already here :cat:

File: lib/hammer_cli_katello/content_view_version.rb [hunk 4]
```diff
@@ -385,6 +385,130 @@ def export_json(content_view_version, repositories)
...
+      end
+
+      def sync_repositories(repositories)
+        repositories.each do |repo|
+          library_repo = index(
+            :repositories,
+            'label' => repo['label'],
+            'library' => true
+          )
+          library_repo = library_repo.first
```
Comment (beav) [inline lib/hammer_cli_katello/content_view_version.rb]: Searching by label just returns everything, the label is ignored.

here's how I worked around it (might not be the most ruby way)

```ruby
          library_repos = index(
            :repositories,
            'content_view_version_id' => 1
          )

          library_repo = library_repos.select do |candidate_repo|
            candidate_repo['label'] == repo['label']
          end

          library_repo = library_repo.first

```

File: lib/hammer_cli_katello/content_view_version.rb [hunk 5]
```diff
@@ -385,6 +385,130 @@ def export_json(content_view_version, repositories)
...
+          )
+        end
+      end
+
+      def repos_units(repositories)
+        repositories.collect do |repo|
+          {
+            'label' => repo['label'],
+            'rpm_filenames' => repo['rpm_filenames'],
+            'errata_ids' => repo['errata_ids']
```
Comment (beav) [inline lib/hammer_cli_katello/content_view_version.rb]: IMO just drop errata_ids for now since it's unused

**Comment by beav** [comment]
@ehelms one comment but I *think* this has everything else. Will need a retest after PR is updated to confirm

File: lib/hammer_cli_katello/content_view_version.rb [hunk 6]
```diff
@@ -385,6 +385,130 @@ def export_json(content_view_version, repositories)
...
+          }
+        end
+      end
+
+      def content_view(name)
+        index(:content_views, 'name' => name).first
```
[the remark this answers is not in the document]
Comment (ehelms) [reply lib/hammer_cli_katello/content_view_version.rb]: I am not sure we know the org ID here. You can import to a different org than you exported from. This is a valid concern though. We may need to have users say import to this org. @beav woudl you agree?

File: test/functional/content_view/version/import_test.rb [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    ex = api_expects(:content_views, :index)
+    ex.returns(
+      'results' => [{
+        'id' => '5',
+        'repositories' => [{'id' => '2', 'label' => 'foo'}],
+        'content_view' => {'name' => 'cv'}
+      }]
+    )
+
+    ex = api_expects(:repositories, :index)
```
[the remark this answers is not in the document]
Comment (ehelms) [reply test/functional/content_view/version/import_test.rb]: I do need a variable, look at the line below. This chains on with_params and then returns.

[the remark this answers is not in the document]
Comment (akofink) [reply test/functional/content_view/version/import_test.rb]: Right, I'm just saying that the convention is to [*actually* chain](https://github.com/Katello/hammer-cli-katello/blob/49b48b3c1d431da10c254e3c59e62552cb909122/test/functional/repository/repository_helpers.rb#L14) `with_params` and `returns`(unless you use a block with `api_expects`). In fact, for repo search you should probably just use that helper.

**Comment by chris1984** [comment]
@ehelms is this good for testing again?

File: lib/hammer_cli_katello/content_view_version.rb [hunk 8]
```diff
@@ -385,6 +385,145 @@ def export_json(content_view_version, repositories)
...
+          if result['option_organization_id'].nil? && result['option_organization_name']
+            result['option_organization_id'] = @command.resolver.organization_id(result)
+          end
+          result
+        end
+      end
+
+      def option_sources
+        super << OrganizationIdParamSource.new(self)
+      end
```
Comment (akofink) [inline lib/hammer_cli_katello/content_view_version.rb]: You don't need to add to option_sources or define a custom option source, afaict. Did you find it necessary for some reason? In addition, you're doing no ID resolution, so I'm not sure why there's logic to support organization name on [L422](https://github.com/Katello/hammer-cli-katello/pull/592/files#diff-c16d36ded485c9e83859c586343e3bb6R422).

File: lib/hammer_cli_katello/content_view_version.rb [hunk 9]
```diff
@@ -385,6 +385,145 @@ def export_json(content_view_version, repositories)
...
+        end
+
+        self.export_tar_file = File.basename(options['option_export_tar'])
+        self.export_tar_prefix = @export_tar_file.gsub('.tar', '')
+        self.export_tar_dir = File.dirname(options['option_export_tar'])
+        untar_export
+
+        export_json = read_json
+
+        cv = content_view(export_json['name'], options['organization_id'])
```
Comment (akofink) [inline lib/hammer_cli_katello/content_view_version.rb]: All the options are prefixed with "option_", so this should be `options['option_organization_id']`. It's currently passing nil to the API.

File: lib/hammer_cli_katello/content_view_version.rb (hunk 8 above)
Comment (ehelms) [reply lib/hammer_cli_katello/content_view_version.rb]: Just following patterns I see in the the code in other places. What is the alternative?

Comment (akofink) [reply lib/hammer_cli_katello/content_view_version.rb]: Remove it all. It works fine without it (try for yourself!).

**Comment by akofink** [comment]
Same thing for export, @ehelms. You can [remove it](https://github.com/Katello/hammer-cli-katello/pull/592/files#diff-c16d36ded485c9e83859c586343e3bb6R287).

**Comment by ehelms** [comment]
Removed

**Comment by ehelms** [comment]
Thanks for the solid reviews!

**Pull request merged**